src/python/pants/core/util_rules/external_tool.py

```diff
@@ -0,0 +1,248 @@
+        ]
+
+        @classmethod
+        def generate_url(cls, plat: Platform, version: str) -> str:
+            ...
+
+        @classmethod
+        def generate_exe(cls, plat: Platform, version: str) -> str:
+            ...
+
+    @rule
+    def my_rule(my_external_tool: MyExternalTool):
+      downloaded_tool = await Get[DownloadedExternalTool](
+        ExternalToolRequest, my_external_tool.get_request(Platform.current)
+      )
+
+    ...
+    def rules():
+      return [my_rule, subsystem_rule(MyExternalTool)]
+
+
+    A lightweight replacement for the code in binary_tool.py and binary_util.py,
+    which can be deprecated in favor of this.
+    """
+
+    # The default values for --version and --known-versions.
+    # Subclasses must set appropriately.
+    default_version: str
+    default_known_versions: List[str]
+
+    @classproperty
+    def name(cls):
+        """The name of the tool, for use in user-facing messages.
+
+        Derived from the classname, but subclasses can override, e.g., with a classproperty.
+        """
+        return cls.__name__.lower()
+
+    @classmethod
+    def get_default_versions_and_digests(cls):
+        """Converts default_known_versions to the old-style default_versions_and_digests.
+
+        Delete this method after --version-digest-mapping below is removed.
+        """
+        # The logic here is a little wonky, because the old default_versions_and_digests format
+        # was itself a little wonky. It could only contain one version+digest per platform,
+        # and the version was presumed to match default_version.
+        # The new default_known_versions can contain digest info for multiple versions, so we
+        # ignore any that don't match default_version.
+        ret = {}
+        for known_version in cls.default_known_versions:
+            try:
+                ver, plat, sha256, length_str = (x.strip() for x in known_version.split("|"))
+            except ValueError:
+                raise ExternalToolError(
+                    f"Bad value for --known-versions (see ./pants "
+                    f"help-advanced {cls.options_scope}): {known_version}"
+                )
+            if ver == cls.default_version:
+                ret[plat] = (ver, sha256, int(length_str))
+        return ret
+
+    @classmethod
+    def register_options(cls, register):
+        super().register_options(register)
+        register(
+            "--version",
+            type=str,
+            default=cls.default_version,
+            advanced=True,
+            fingerprint=True,
+            help=f"Use this version of {cls.name}.",
+        )
+        register(
+            "--known-versions",
+            type=list,
+            member_type=str,
+            default=cls.default_known_versions,
+            advanced=True,
+            help=f"Known versions to verify downloads against. Each element is a "
```

Comment:
Would something like:
```python
      dedent(f"""\
      ...
      """)
```
Potentially be easier to read and change?

Reply:
Sorry I didn't address these comments Danny. I somehow missed them! Will look now and fix in my current change as needed.

Re this, I'm thinking the rendering code should figure this sort of thing out, as needed?

Reply:
Oh sure! I meant for the purposes of development, the result might look cleaner. I agree that the rendering code should be figuring out how to display the text.